### PR TITLE
fix: persist rotated health-check tokens

### PR DESCRIPTION
**Pull request by lubshad**
## Summary

- Persist rotated, single-use OAuth refresh tokens before reporting account health.
- Share transactional refresh persistence across health, refresh, and doctor tools.
- Keep doctor diagnostics and auto-switching correct when refresh results or storage change concurrently.

## Testing

- [x] `npm run typecheck`
- [x] `npm run lint`
- [x] `npm test`
- [x] `npm run build`

## Compliance Confirmation

- [x] This change stays within the repository scope and OpenAI Terms of Service expectations.
- [x] This change uses official authentication flows only and does not add bypass, scraping, or credential-sharing behavior.
- [x] I updated tests and documentation when the change affected users, maintainers, or repository behavior.

## Notes

- Linked issue: none.
- Follow-up work or rollout notes: includes regressions for token rotation persistence, concurrent re-login protection, doctor mixed verification results, and storage reordering.\n\n
## Summary by CodeRabbit

- **New Features**
  - Added a refresh-verification-failure indicator that blocks affected accounts and triggers an immediate re-authentication next action.
  - Health/refresh reporting now distinguishes refreshed, skipped (disabled), and failed accounts, including skipped-count reporting in JSON/UI.

- **Bug Fixes**
  - Improved refresh-token rotation persistence across related workspace records while preserving concurrent updates during repair/fix flows.
  - Doctor/health/refresh now reload cached account-manager state after refresh so results match live outcomes.

- **Tests**
  - Expanded transactional persistence and refresh-verification edge-case coverage; refined concurrency and exact error-message assertions.

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

this pr persists rotated oauth refresh tokens before reporting account health. the main changes are:

- shared transactional refresh persistence across health, refresh, and doctor tools.
- cache reloads after token storage changes.
- refreshed doctor diagnostics and identity-based auto-switching.
- expanded coverage for token safety and concurrent storage changes.

<h3>Confidence Score: 5/5</h3>

the latest fixes look safe to merge.

- concurrent re-login credentials are preserved.
- doctor results are rebuilt from current storage.
- account switching no longer writes a stale array index.
- no additional blocking issue qualifies for this follow-up round.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/tools/refresh-account.ts | adds shared transactional refresh persistence with concurrent credential-change protection. |
| lib/tools/codex-doctor.ts | reloads diagnostics after verification and resolves auto-switch targets by identity. |
| lib/tools/codex-health.ts | persists rotated tokens before reporting health and reports disabled accounts separately. |
| lib/tools/codex-refresh.ts | moves manual refreshes onto the shared persistence path. |
| index.ts | adds cache reload logic that flushes pending saves before reading updated storage. |

<sub>Reviews (4): Last reviewed commit: ["fix: reuse safe account manager reload"](https://github.com/ndycode/oc-codex-multi-auth/commit/9e04513b6d509515d4bf1ffc263535f67c99fe25) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=46575136)</sub>

**Context used:**

- Context used - speak in lowercase, concise sentences. act like th... ([source](https://app.greptile.com/review/custom-context?memory=instruction-0))

### CODE REVIEW

File: lib/tools/refresh-account.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+					throw new Error(
+						"Refresh token changed concurrently while verification was in progress",
+					);
+				}
+				target.refreshToken = refreshResult.refresh;
+				target.accessToken = refreshResult.access;
+				target.expiresAt = refreshResult.expires;
+				if (rotatedAt !== undefined) {
+					target.tokenRotatedAt = rotatedAt;
+				}
```
[the remark this answers is not in the document]
Comment (lubshad) [reply lib/tools/refresh-account.ts]: Fixed in 29748a3. Removed the sibling replacement-count fallback, so a target whose credential changed concurrently now fails closed and preserves the newer login. Added a regression test for this race.

File: lib/tools/codex-doctor.ts [hunk 2]
```diff
@@ -304,7 +342,19 @@ export function createCodexDoctorTool(ctx: ToolContext): ToolDefinition {
...
+				// The initial diagnostics snapshot was taken before token verification.
+				// Reload after fixes so the reported health never contradicts live
+				// refresh results (e.g. "8 healthy" alongside eight invalid tokens).
+				diagnostics = await loadDiagnostics(extraFindings);
+				if (verificationFailures > 0 && diagnostics.summary.healthy > 0) {
+					diagnostics.summary.healthy = Math.max(
+						0,
+						diagnostics.summary.healthy - verificationFailures,
+					);
+				}
```
[the remark this answers is not in the document]
Comment (lubshad) [reply lib/tools/codex-doctor.ts]: Fixed in 29748a3. Doctor now marks failed verification identities unavailable before rebuilding summary, findings, and next action. Added mixed success/failure coverage.

File: lib/tools/codex-doctor.ts [hunk 3]
```diff
@@ -272,15 +303,22 @@ export function createCodexDoctorTool(ctx: ToolContext): ToolDefinition {
...
+								const currentActive = resolveActiveIndex(current, "codex");
+								if (best.index === currentActive) return false;
+								current.activeIndex = best.index;
+								current.activeIndexByFamily =
+									current.activeIndexByFamily ?? {};
+								for (const family of MODEL_FAMILIES) {
+									current.activeIndexByFamily[family] = best.index;
+								}
+								await persist(current);
+								return true;
```
[the remark this answers is not in the document]
Comment (lubshad) [reply lib/tools/codex-doctor.ts]: Fixed in 29748a3. Auto-switch now resolves the selected account by stable identity inside the storage transaction instead of persisting the stale manager index. Added a storage-reorder regression test.

File: lib/tools/codex-doctor.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lubshad) [reply lib/tools/codex-doctor.ts]: Fixed in 29748a3. Verification failures now enter diagnostic snapshots before derived summary, findings, and next action are calculated; the post-build healthy-count adjustment was removed.

**Comment by lubshad** [comment]
Updated in 9e04513:

- reused the guarded cached-manager reload in the account-selection event path
- documented and asserted the doctor auto-switch transaction sequence
- added reload ordering, disposal, failure-retention, and retry coverage

Validation: `npm run typecheck`, `npm run lint`, and `npm test` (2,823 tests) all pass.
